File: pkg/server/client/camera.go
package client

import (
	"math"

	"github.com/namelessmmo/realm/pkg/server/location"
)

type Camera struct {
	screenWidth  int
	screenHeight int

	Location *location.Location
}

func NewCamera(screenWidth, screenHeight int) *Camera {
	return &Camera{
		screenWidth:  screenWidth,
		screenHeight: screenHeight,
	}
}

func (camera *Camera) update(loc *location.Location) {
	world := loc.GetWorld()
	centerX := camera.screenWidth / 2
	centerY := camera.screenHeight / 2

	cameraX := loc.GetX() - centerX
	cameraY := loc.GetY() - centerY
	cameraX = int(math.Max(0, math.Min(float64(cameraX), float64(world.GetWidth()-camera.screenWidth))))
	cameraY = int(math.Max(0, math.Min(float64(cameraY), float64(world.GetHeight()-camera.screenHeight))))

	camera.Location = location.NewLocation(loc.GetWorld(), cameraX, cameraY)
}

File: public/src/game/camera.ts
import { Location } from "./location";

export class Camera {
    public x: number;
    public y: number;
    public readonly screenWidth: number;
    public readonly screenHeight: number;

    constructor(screenWidth: number, screenHeight: number) {
        this.screenWidth = screenWidth;
        this.screenHeight = screenHeight;
    }

    public update(location: Location) {
        const centerX = this.screenWidth / 2;
        const centerY = this.screenHeight / 2;

        let cameraX = location.x - centerX;
        let cameraY = location.y - centerY;
        cameraX = Math.max(0, Math.min(cameraX, (location.world.tilemap.width - this.screenWidth)));
        cameraY = Math.max(0, Math.min(cameraY, (location.world.tilemap.height - this.screenHeight)));

        this.x = cameraX;
        this.y = cameraY;
    }
}
